### ISSUE TITLE
ClassCastException in Rubric questions

### ISSUE THREAD

**Issue by damithc** [TEAMMATES/repo#4319]
Trying to convert a Text to String?
```
teammates.common.datatransfer.FeedbackTextResponseDetails cannot be cast to teammates.common.datatransfer.FeedbackRubricResponseDetails


Request Path /page
Request Parameters {...}
Stack Trace



java.lang.ClassCastException: teammates.common.datatransfer.FeedbackTextResponseDetails cannot be cast to teammates.common.datatransfer.FeedbackRubricResponseDetails
	at teammates.common.datatransfer.FeedbackRubricQuestionDetails.calculateResponseFrequency(FeedbackRubricQuestionDetails.java:562)
	at teammates.common.datatransfer.FeedbackRubricQuestionDetails.getQuestionResultStatisticsHtml(FeedbackRubricQuestionDetails.java:456)
	at teammates.ui.controller.InstructorFeedbackResultsPageData.buildQuestionTableAndResponseRows(InstructorFeedbackResultsPageData.java:1055)
	at teammates.ui.controller.InstructorFeedbackResultsPageData.buildQuestionTableAndResponseRows(InstructorFeedbackResultsPageData.java:991)
	at teammates.ui.controller.InstructorFeedbackResultsPageData.initForViewByQuestion(InstructorFeedbackResultsPageData.java:153)
	at teammates.ui.controller.InstructorFeedbackResultsPageAction.execute(InstructorFeedbackResultsPageAction.java:132)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:337)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:38)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:142)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:35)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:257)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:442)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:230)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:234)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by LowWeiLin** [comment]
There seems to be a text question response saved for a rubric question...

**Comment by damithc** [comment]
Raising to critical. An instructor cannot access session results because of this error.

**Comment by damithc** [comment]
How could that happen? I tried a few older rubric questions and they seem to work fine.

**Comment by LowWeiLin** [comment]
I'm not really sure. Is it possible to check the responses for the question to see if there really is a text response?

**Comment by damithc** [comment]
Seems all responses for that question ID are RUBRIC type 
![image](https://cloud.githubusercontent.com/assets/1673303/9927243/7feff2e6-5d4f-11e5-8052-254e2ca63fd4.png)

**Comment by damithc** [comment]
Here's the question: 
![image](https://cloud.githubusercontent.com/assets/1673303/9927299/155812e6-5d50-11e5-9685-8c179d801703.png)

**Comment by damithc** [comment]
Perhaps add more logging in a branch? I'll deploy that branch and try to access the same page to figure out which exact data item is causing this problem

**Comment by damithc** [comment]
Is this how it is supposed to be?
![image](https://cloud.githubusercontent.com/assets/1673303/9927351/874cb546-5d50-11e5-9110-d628dd388c00.png)

**Comment by LowWeiLin** [comment]
Yes that seems fine. "answer" is a text type containing json.
Will add more logging, and ignore the offending response so that the results can be seen.

**Comment by damithc** [comment]
I opened the 'all records' page of each students in that course. They all worked. I was able to see the responses to the rubric question :-p

**Comment by kanghj** [comment]
Was that any sort of task queue operation that affected the session? e.g. submissions adjustment. Or the instructor modified question numbers in InstructorFeedbackEdit?

**Comment by damithc** [comment]
He has deleted a few students. From the recent logs, I can't see any QuestionEdit actions.

**Comment by damithc** [comment]
Sorry, deletes were in other courses. No student deletes in this particular course.

**Comment by damithc** [comment]
BTW, I managed to rectify the situation by editing question numbers. But we still need to know how the question numbers became all 1. Keep in mind that relying on the database to find the next index is not reliable due to eventual consistency.

**Comment by damithc** [comment]
Perhaps we can put in some measures to detect it and report it (for manual correction) or auto-correct if possible. Also make the code more resilient to such errors e.g. group responses by question id rather than question number.

**Comment by kanghj** [comment]
I think this bug could be more complicated that that. The responses are already grouped by question, and depends on FeedbackQuestionAttributes <code>.equals</code> which tests for more than just the question number. 

When I edit a question's number on the staging server so that two questions (with different question types) have the same question number, the behaviour is different from this bug.

**Comment by damithc** [comment]
Try making all questions have number 1 and also include a rubric question. One other thing: the session edit page showed correct question numbers although database has '1' for all the questions. When I changed the numbers in the database, the session went back to normal.

**Comment by kanghj** [comment]
The edit page shows numbers that look correct because the javascript on that page sets the number based on the ordering of the questions on the page. 

Managed to replicate the bug on dev server. Will make a new issue and make a pr.

**Comment by LowWeiLin** [comment]
Closing this issue, to be fixed in #4321 and #4328.
